File: python-code/modules/html_in_png.py
#!/usr/bin/env python3

'''
Creates an HTML+png polyglot for Internet Explorer
'''

import argparse
import re
import random
import pngdata

RANDOM_DATA_SIZE = 1900

STEGO = """
                                  .       .
                                 / `.   .' \\
                         .---.  <    > <    >  .---.
                         |    \\  \\ - ~ ~ - /  /    |
                          ~-..-~             ~-..-~
                      \\~~~\\.'   stegosploit      `./~~~/
                       \\__/                imajs   \\__/
                         /  .    -.                 \\
                  \\~~~\\/   {       \\       .-~ ~-.    -._ _._
                   \\__/    :        }     {       \\     ~{  p'-._
    .     .   |~~. .'       \\       }      \\       )      \\  ~--'}
    \\\\   //   `-..'       _ -\\      |      _\\      )__.-~~'='''''
 `-._\\\\_//__..-'      .-~    |\\.    }~~--~~  \\=   / \\
  ``~---.._ _ _ . - ~        ) /=  /         /+  /   }
                            /_/   /         {   } |  |
                              `~---o.       `~___o.---'
"""


def genRandomHTMLChar():
	while True:
		x = random.randint(0,255)
		if x != 0 and x!=ord('<') and x!=ord('>') and x!=ord('/'):
			return x

def main():
	print(STEGO)	# print the stego

	description = 'Creates and HTML+png polyglot for Internet Explorer.'
	parser = argparse.ArgumentParser(description = description)
	parser.add_argument('htmlfile', 
			type=argparse.FileType('r'),
			nargs='?',					# REMOVE ME: sets the arg optional
			help='HTML file name')

	parser.add_argument('pngfile', 
			type=argparse.FileType('rb'),
			nargs='?',					# REMOVE ME
			help='PNG file name')

	parser.add_argument('output', 
			type=argparse.FileType('wb'),
			nargs='?',					# REMOVE ME
			help='Output file name')
	args = parser.parse_args()

	htmlData = ''		
	if args.htmlfile:
		htmlData = args.htmlfile.read()
		htmlData = htmlData.rstrip()	# remove trailing whitespace

	# pngData = open('anon.png','rb').read()	# FIXME
	if args.pngfile:
		png_data = args.pngfile.read()
		# print(png_data)

	open_comment = pngdata.make_text_chunk("<html>", "<!-- ")
	png_data_array = list(png_data)
	png_data_array[2:2] = open_comment

	# pngStart = pngData[:4]		# read the first 4 bytes (Magic # = 0xFFD8FFE0)
	# jfifapp0 = pngData[6:20]	# from JFIF to offset 20
	# restOfpng = pngData[20:]	# read the rest of the file

	# pngData = None	# remove pngData from cache for performance

	# print('',pngStart,'\n',jfifapp0,'\n',restOfpng)

	# html = '<html><!--'
	# content = ' -->'

	# remove <html>,<head> and <meta http-equiv ...> tags
	# from the html
	htmlTags = re.compile('<html>')
	headTags = re.compile('<head>')
	metaTags = re.compile('<meta http-equiv[^>]*>')

	htmlData = re.sub(htmlTags,'',htmlData)

	htmlData = " -->" + htmlData +  "<script type='text/undefined'>/*"

	# htmlData = re.sub(headTags,'',htmlData)
	# htmlData = re.sub(metaTags,'',htmlData)

	# content += htmlData + '<!--'

	# paddingLength= 0x2f2a - len(content) - len(html)
	paddingLength = random.randint(0,RANDOM_DATA_SIZE)

	# 	padding = ''
	for i in range(paddingLength):
		x = genRandomHTMLChar()
		htmlData = chr(x) + htmlData

	# prePadding = padding[0:randomLength]
	# postPadding = padding[randomLength:]
	
	html_Content = pngdata.make_text_chunk("_", htmlData)
	png_data_array[3:3] = html_Content

	# lots of print statements
	# print('HTML header length = %d' % len(html))
	# print('Pre padding length = %d' % len(prePadding))
	# print('Post padding length = %d' % len(postPadding))
	# print('HTML data length = %d' % len(htmlData))
	# print('Content length = %d' % len(content))
	# print('Final content length = %d' % len(finalContent))

	# newpng = ''.join(png_data_array)
	print(png_data_array)
	newpng = bytes(png_data_array)
	# pp

	# print(newpng)		# Output to file
	
	with open(args.output, 'wb') as f: 
		f.write(newpng) 

if __name__ == '__main__':
	main()

File: imajs/html_in_png.pl
#!/usr/bin/perl
#
#                                  .       .
#                                 / `.   .' \
#                         .---.  <    > <    >  .---.
#                         |    \  \ - ~ ~ - /  /    |
#                          ~-..-~             ~-..-~
#                      \~~~\.'   stegosploit      `./~~~/
#                       \__/                        \__/
#                         /  .    -.                  \
#                  \~~~\/   {       \       .-~ ~-.    -._ _._
#                   \__/    :        }     {       \     ~{  p'-._
#    .     .   |~~. .'       \       }      \       )      \  ~--'}
#    \\   //   `-..'       _ -\      |      _\      )__.-~~'='''''
# `-._\\_//__..-'      .-~    |\.    }~~--~~  \=   / \
#  ``~---.._ _ _ . - ~        ) /=  /         /+  /   }
#                            /_/   /         {   } |  |
#                              `~---o.       `~___o.---'
#
#
# by Saumil Shah @therealsaumil
#
# This script creates an HTML+PNG polyglot
#
# Requires PNGDATA.pm, CRC32.pm

use FindBin;
use lib $FindBin::Bin;
use PNGDATA;

$| = 1;

$RANDOM_DATA_SIZE = 1900;

$htmlfile = $ARGV[0];
$pngfile = $ARGV[1];
$outfile = $ARGV[2];

print <<"EOT";

                                  .       .
                                 / `.   .' \\
                         .---.  <    > <    >  .---.
                         |    \\  \\ - ~ ~ - /  /    |
                          ~-..-~             ~-..-~
                      \\~~~\\.'   stegosploit      `./~~~/
                       \\__/                imajs   \\__/
                         /  .    -.                 \\
                  \\~~~\\/   {       \       .-~ ~-.    -._ _._
                   \\__/    :        }     {       \\     ~{  p'-._
    .     .   |~~. .'       \\       }      \\       )      \\  ~--'}
    \\\\   //   `-..'       _ -\\      |      _\\      )__.-~~'='''''
 `-._\\\\_//__..-'      .-~    |\\.    }~~--~~  \\=   / \\
  ``~---.._ _ _ . - ~        ) /=  /         /+  /   }
                            /_/   /         {   } |  |
                              `~---o.       `~___o.---'

EOT

if($#ARGV < 2) {
   print "Usage: $0 <html source> <png source> <output>\n";
   exit;
}

local $/;

$htmldata = "";

if($htmlfile ne '-') {
   open(HTMLFILE, $htmlfile) || die("Cannot open $htmlfile\n");
   $htmldata = <HTMLFILE>;
   close(HTMLFILE);
}

$htmldata =~ s/[\r\n]+$//;    # remove trailing newline characters

open(PNGFILE, $pngfile) || die("Cannot open $pngfile\n");
$pngdata = <PNGFILE>;
close(PNGFILE);

my @png = ();

@png = PNGDATA::read_png($pngdata);

# insert a tEXt chunk with an HTML comment
# after the PNG header and the IHDR chunk
# i.e. as the 3rd array element
$open_comment = PNGDATA::make_text_chunk("<html>", "<!-- ");
splice(@png, 2, 0, $open_comment);
$htmldata =~ s/<html>//i;
#$htmldata .= "<!-- <![CDATA[";
$htmldata = " -->" . $htmldata . "<script type='text/undefined'>/*";   # works best in blocking the rest of the data

# add some random data at the end of the HTML data

$padding_length = int(rand($RANDOM_DATA_SIZE));

print "Appending $padding_length random bytes\n";

for($i = 0; $i < $padding_length; $i++) {
   while(1) {
      $x = int(rand(255));
      if($x >= 32 && $x != 0 && $x != ord('<') && $x != ord('>') && $x != ord('/')) {
         last;
      }
   }
   $htmldata = chr($x) . $htmldata;
}

$html_content = PNGDATA::make_text_chunk("_", $htmldata);
splice(@png, 3, 0, $html_content);

PNGDATA::print_png_data(@png);

$newpng = join('', @png);

open(OUTPUT, ">$outfile") || die("Cannot write to $outfile\n");
print OUTPUT $newpng;
close(OUTPUT);

print "Written $outfile\n";

